apps/web/src/dialogs/command-palette/command-palette-dialog.tsx

```diff
@@ -0,0 +1,472 @@
+/*
+This file is part of the Notesnook project (https://notesnook.com/)
+
+Copyright (C) 2023 Streetwriters (Private) Limited
+
+This program is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+import { debounce, toTitleCase } from "@notesnook/common";
+import { ScrollContainer } from "@notesnook/ui";
+import { Button, Flex, Text } from "@theme-ui/components";
+import { useEffect, useRef, useState } from "react";
+import { db } from "../../common/db";
+import { BaseDialogProps, DialogManager } from "../../common/dialog-manager";
+import Dialog from "../../components/dialog";
+import Field from "../../components/field";
+import {
+  Cross,
+  Icon,
+  Note as NoteIcon,
+  Notebook as NotebookIcon,
+  Reminder as ReminderIcon,
+  Tag as TagIcon
+} from "../../components/icons";
+import { hashNavigate, navigate } from "../../navigation";
+import { useEditorStore } from "../../stores/editor-store";
+import Config from "../../utils/config";
+import { commands as COMMANDS } from "./commands";
+
+interface Command {
+  id: string;
+  title: string;
+  type: "command" | "note" | "notebook" | "tag" | "reminder";
+  group: string;
+}
+
+type GroupedCommands = Record<
+  string,
+  (Command & { index: number; icon: Icon | undefined })[]
+>;
+
+export const CommandPaletteDialog = DialogManager.register(
+  function CommandPaletteDialog(props: BaseDialogProps<boolean>) {
+    const [commands, setCommands] = useState<Command[]>(getDefaultCommands());
+    const [selected, setSelected] = useState(0);
+    const [query, setQuery] = useState(">");
+    const selectedRef = useRef<HTMLButtonElement>(null);
+
+    useEffect(() => {
+      selectedRef.current?.scrollIntoView({
+        block: "nearest"
+      });
+    }, [selected]);
+
+    function reset() {
+      setSelected(0);
+      setQuery(">");
+      setCommands(getDefaultCommands());
+    }
+
+    const highlighted = db.lookup.fuzzy(
+      prepareQuery(query),
+      commands,
+      "title",
+      {
+        prefix: "<b style='color: var(--accent-foreground)'>",
+        suffix: "</b>"
+      }
+    );
+    const grouped = highlighted.reduce((acc, command, index) => {
+      if (!acc[command.group]) {
+        acc[command.group] = [];
+      }
+      acc[command.group].push({
+        ...command,
+        icon: getCommandIcon(command),
+        index
+      });
+      return acc;
+    }, {} as GroupedCommands);
```

Comment:
Why is it done here instead of in the `onChange` function?

Reply:
I think groupifying the commands is only needed during render, that's why it is here. Do you mean we should have `grouped` be managed in state, or do you have something else in mind?